photon-ml/src/main/scala/com/linkedin/photon/ml/FeatureIndexingJob.scala

```diff
@@ -138,39 +171,159 @@ class FeatureIndexingJob(val sc: SparkContext,
   }
 
   def run(): Unit = {
-    val inputRdd = AvroIOUtils.readFromAvro[GenericRecord](sc, inputPath, 10)
-    val featuresRdd = partitionedUniqueFeatures(inputRdd)
-    buildIndexMap(featuresRdd)
+    val inputRdd = AvroUtils.readAvroFiles(sc, inputPaths, 10)
+
+    ensureOutputPath(outputPath)
+
+    featureShardIdToFeatureSectionKeysMap match {
+      case Some(shardToSectionsMap) =>
+        // If a shard to feature section set was specified, build an index for each shard
+        shardToSectionsMap.map { case (shardId, featureSections) =>
+          val featuresRdd = partitionedUniqueFeatures(inputRdd, Some(featureSections))
+          buildIndexMap(featuresRdd, outputPath, shardId)
+        }
+
+      case _ => {
+        // Otherwise, build a global index
+        val featuresRdd = partitionedUniqueFeatures(inputRdd)
```

Comment:
Interesting. This would break the nice conceptual mapping we have between feature shards and the individual feature vectors in the GAME model though, wouldn't it?